### ISSUE TITLE
update imgadm to use newer docker-registry-client so it can pull from v2?

### ISSUE THREAD

**Issue by h0lzi** [joyent/smartos-live#644]
would be nice if imgadm could be updated to use newer node-docker-registry-client so that it can pull from v2 docker registry. it probably needs changes in src/img/lib/sources/docker.js too.

Reason:
so people that do "docker" on standalone smartos (e.g. http://www.cruwe.de/2016/01/27/using-docker-on-smartos-hypervisors.html) can have v2 registry pulls.

**Comment by christophkreutzer** [comment]
Hi @trentm,

I've seen there is some progress in [RFD 57](https://github.com/joyent/rfd/tree/master/rfd/0057). Is there any way we can support you?
I'm fairly experienced with nodejs and would contribute, if you could point me a direction. However, I'm not that firm with imgadm ;-)

**Comment by trentm** [comment]
@christophkreutzer RFD 57 work is in testing right now... perhaps once that is merged to master (hopefully that is within a couple of weeks)?

**Comment by trentm** [comment]
BTW the primary internal ticket for RFD 57 work is: https://smartos.org/bugview/DOCKER-929

**Comment by christophkreutzer** [comment]
Ok, thanks! Just keep me posted if there is something to do.

**Comment by twhiteman** [comment]
@trentm I've updated the code to use docker v2 manifests, but there are some new node_modules entries that I wanted to bring up first.

I've managed to reduce the node_modules changes significantly by making docker-registry-client use the older 'restify' module instead of 'restify-clients' and 'restify-errors' - otherwise these two new restify modules would bring in a huge amount of additional dependencies (i.e. ending up with a ~10MB increase). As it is this increases the total size by ~1MB (to 2.7MB), and half of that increase comes from the new 'jwk-to-pem' module.

Note: If/when the docker-registry-client drops support for v1 and the old v2.1 docker manifest formats, that space would be reclaimed (dropping most of these added modules, e.g. 'jwk-to-pem').

Here is the summary of node_module changes (the last ?? lines will become newly added modules):
```bash
$ git status --short
 M node_modules/docker-registry-client/lib/common.js
AM node_modules/docker-registry-client/lib/docker-json-client.js
A  node_modules/docker-registry-client/lib/errors.js
 M node_modules/docker-registry-client/lib/index.js
 M node_modules/docker-registry-client/lib/registry-client-v1.js
AM node_modules/docker-registry-client/lib/registry-client-v2.js
 M node_modules/docker-registry-client/node_modules/tough-cookie/lib/cookie.js
 M node_modules/docker-registry-client/node_modules/tough-cookie/lib/memstore.js
A  node_modules/docker-registry-client/node_modules/tough-cookie/lib/pathMatch.js
A  node_modules/docker-registry-client/node_modules/tough-cookie/lib/permuteDomain.js
 M node_modules/docker-registry-client/node_modules/tough-cookie/lib/pubsuffix.js
 M node_modules/docker-registry-client/node_modules/tough-cookie/lib/store.js
 D node_modules/docker-registry-client/node_modules/tough-cookie/node_modules/punycode/package.json
 D node_modules/docker-registry-client/node_modules/tough-cookie/node_modules/punycode/punycode.js
 M node_modules/docker-registry-client/node_modules/tough-cookie/package.json
 M node_modules/docker-registry-client/package.json
 M node_modules/imgmanifest/lib/imgmanifest.js
?? node_modules/concat-stream/
?? node_modules/docker-registry-client/node_modules/base64url/
?? node_modules/docker-registry-client/node_modules/jwk-to-pem/
?? node_modules/docker-registry-client/node_modules/jws/
?? node_modules/docker-registry-client/node_modules/www-authenticate/
```

**Comment by DvdGiessen** [comment]
I've just spend some time testing the changes from the gh-644 branch and Docker v2 manifest support works fine: I'm able to import a Docker image using the v2 manifest format and use it to create a LX zone.

However, images only available via the v1 manifest format are no longer working? Specifically, in [`src/img/lib/sources/docker.js:150`](https://github.com/joyent/smartos-live/commit/403d13a8371d15c0b80b7b1eabe79626cb01828e#diff-9cd349fa5efd708f61fd326a4a25e4b6R150) there is a check requiring manifests to be for version 2.  The test suite also does not complete successfully due to this change.

**Comment by sjuxax** [comment]
@trentm @twhiteman Just curious, what specifically is holding this up now? 
Is there something specific that is wrong with the existing branch that could potentially be patched? The Joyent Gerrit entry at https://cr.joyent.us/#/c/2645 seems to indicate that this changeset is still pending review and integration approval, and the internal DOCKER-929 bug was marked as finished in July.

It'd be great if I could easily pull images from Docker Hub onto a SmartOS install, and obviously a big help in convincing people that SmartOS is a viable cloud OS. If I have to install custom patches to easily run Docker images, that goes right the window. 😛 I understand that this functionality is already implemented on Triton, but it's a lot more work to get local Triton configured for experimentation than SmartOS.

**Comment by twhiteman** [comment]
This initially stalled on design discussions about how to include the docker bits with imgadm (if at all) - such as moving to a plugin, or moving to a separate component/binary - as you shouldn't need to update/reboot machines to support a newer version of the imgadm docker client...

Anyway, the imgadm docker code has been updated and I'll try to push this along as is - as I don't think anyone uses v1 registries anymore... something working is better than nothing working.

**Comment by drscream** [comment]
@twhiteman I agree, I also assume nobody is using v1 anymore. I've created a small blog post about the docker integration based on your last patch which I merged into our SmartOS build: https://www.cyber-tec.org/2018/02/11/run-docker-images-on-smartos/

I will check your last changes and try to do a new build in the upcoming week.

**Comment by twhiteman** [comment]
Now committed in afe9efb0.

**Comment by sjorge** [comment]
Awesome! Looking forward to playing with this in the next release
